### PR TITLE
[AutoScheduler] Add winograd support in tuning networks

### PR DESCRIPTION
**Pull request by merrymercy**
This PR adds the Winograd support for NHWC layout.

Now it is only enabled for GPU backend, but the CPU backend uses the same compute declaration. I will enable it in later PRs.

cc @jcf94 @FrozenGene @comaniac

### CODE REVIEW

File: python/tvm/relay/op/strategy/cuda.py [hunk 1]
```diff
@@ -206,19 +198,39 @@ def conv2d_strategy_cuda(attrs, inputs, out_type, target):
...
+            ):
+                strategy.add_implementation(
+                    wrap_compute_conv2d(topi.cuda.conv2d_nhwc_tensorcore),
+                    wrap_topi_schedule(topi.cuda.schedule_conv2d_nhwc_tensorcore),
+                    name="conv2d_nhwc_tensorcore.cuda",
+                    plevel=20,
+                )
+
+            # register auto-scheduler implementations
+            judge_winograd_auto_scheduler = (
```
Comment (comaniac) [inline python/tvm/relay/op/strategy/cuda.py:218]: Following the above style of judging Winograd, we can remove this variable.

File: python/tvm/relay/op/strategy/cuda.py [hunk 2]
```diff
@@ -206,19 +198,39 @@ def conv2d_strategy_cuda(attrs, inputs, out_type, target):
...
+                ("float" in data.dtype and "float" in kernel.dtype)
```
Comment (comaniac) [inline python/tvm/relay/op/strategy/cuda.py:219]: Will `dtype` be "float32"?

File: src/auto_scheduler/search_policy/sketch_policy.cc [hunk 3]
```diff
@@ -116,8 +116,8 @@ SketchPolicy::SketchPolicy(SearchTask task, CostModel program_cost_model,
   } else if (IsGPUTask(node->search_task)) {
     // Sketch Generation Rules
     node->sketch_rules.push_back(&rule_add_cache_read_stage);
-    node->sketch_rules.push_back(&rule_always_inline);
     node->sketch_rules.push_back(&rule_special_compute_location_gpu);
+    node->sketch_rules.push_back(&rule_always_inline);
```
Comment (comaniac) [inline src/auto_scheduler/search_policy/sketch_policy.cc:120]: Out of curiosity, why changed to this order? I saw that the condition of applying special compute location also checks should be always inlined, so then it should be fine to keep always inline to go first?

**Review by comaniac — CHANGES_REQUESTED** [review]

File: python/tvm/topi/nn/conv2d.py [hunk 4]
```diff
@@ -943,3 +943,212 @@ def unpack_NCHWc_to_nchw(packed_out, out_dtype):
...
+    # transform data
+    r_a = te.reduce_axis((0, alpha), "r_a")
+    r_b = te.reduce_axis((0, alpha), "r_b")
+    data_pack = te.compute(
+        (alpha, alpha, P, CI),
+        lambda eps, nu, p, ci: te.sum(
+            input_tile[r_a][r_b][p][ci] * B[r_a][eps] * B[r_b][nu], axis=[r_a, r_b]
+        ),
+        name="data_pack",
+        attrs={"auto_scheduler_simplify_const_tensor_indices": ["eps", "nu", "r_a", "r_b"]},
```
Comment (FrozenGene) [inline python/tvm/topi/nn/conv2d.py:1048]: We could add one comment what the attribute is to do and why we need this for auto scheduler

File: python/tvm/relay/op/strategy/cuda.py (hunk 2 above)
Comment (FrozenGene) [reply python/tvm/relay/op/strategy/cuda.py:219]: float16 should work too.

**Review by FrozenGene — APPROVED** [review]

File: python/tvm/relay/op/strategy/cuda.py (hunk 2 above)
Comment (comaniac) [reply python/tvm/relay/op/strategy/cuda.py:219]: nvm, I was thinking in a wrong way.

File: src/auto_scheduler/search_policy/sketch_policy.cc (hunk 3 above)
Comment (jcf94) [reply src/auto_scheduler/search_policy/sketch_policy.cc:120]: I have the same opinion as Cody here.

Comment (merrymercy) [reply src/auto_scheduler/search_policy/sketch_policy.cc:120]: The old order generates wrong sketches for winograd.

**Pull request merged**